Enhance: allow modification of the player room indication on 2D map

SlySven opened the pull request:
This commit reorganises the 2D Painter code to draw the player room indicator AFTER all the rooms have been drawn so that it does not get overdrawn by rooms painted after the player's room - which it could be if the rooms are at maximum size or the area is in "gridMode".

It also offers some alternative marker options in the form of an oversized ring around the room which reduces the obstruction of the details of the room when the marker is drawn on it. The alternatives are:
* a fixed red colour ring
* a fixed blue/yellow colour ring - two contrasting colours mean that it can not be lost if drawn over a room that already has one of them as its colour (environment) setting.
* a customizable two colour ring where the user can set the outer and inner colours as they like - even to the same colour.

~This commit does not contain code to save the setting between sessions. It is intended to get peer/public evaluation of the proposal.~ A follow up commit ~will be~ *has been* appended ~if~ *as* this proposal is acceptable and does work.

The ancient `(bool) Host::mMapStrongHighlight` option is still respected, I note that it was introduced in 3cb1719b0f516eae4fdbd9288e3f8e4e83e9ccbf from 2011/01/20 (after Mudlet 1.0.5) and the code to control it was lost in 8b0e8bb0f925d58c558cf864faae6b987435bd82 from 2012/12/31 "NEW Vadim Peretokin: mapper remembers settings" just before the release of Mudlet 2.0...!

This may fix #2057 raised by @Eraene .

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven commented:
Original, current code:
![Screenshot_20190615_165322](https://user-images.githubusercontent.com/6163092/59554838-c04c9480-8fa9-11e9-9d13-028545b5f4f6.png)

Emulation of original - however drawing the indication later in the code means the exit markings within the room have been covered - to get it back we would have to have a copy of the code for them as part of the processing of the player room (making a function to do that and calling that for every room normally will slow things unless it is made an `inline` one):
![Screenshot_20190615_160546](https://user-images.githubusercontent.com/6163092/59554842-c0e52b00-8fa9-11e9-8531-4860359ad9d8.png)

Red ring:
![Screenshot_20190615_160630](https://user-images.githubusercontent.com/6163092/59554841-c0e52b00-8fa9-11e9-9726-3a9fade8b0a2.png)

Blue/Yellow Ring:
![Screenshot_20190615_160710](https://user-images.githubusercontent.com/6163092/59554840-c0e52b00-8fa9-11e9-95c8-f0a508e8a37d.png)

Custom ring with green and magenta colours:
![Screenshot_20190615_160821](https://user-images.githubusercontent.com/6163092/59554839-c0e52b00-8fa9-11e9-827a-c498c453fb79.png)

vadi2 commented:
I have a different take on doing this - what if we allow people to set a custom image in place of the marker? That would be way more flexible than the current and proposed ring system.

Eraene commented:
This seems like a good change. I'm a little concerned the ring is too large, as it covers up adjacent rooms just slightly (and their exits), but it's a large step in the right direction.

SlySven commented:
Also you would also need an image with a degree of transparency - and I do not imaging that there is a large proportion of stock images with that characteristic - remember `.jpg` just doesn't handle transparently and `.gif` can only set pixels of particular pixel colour to wholly transparent so you are pretty much talking only of `.png` ...

SlySven commented:
> ... I'm a little concerned the ring is too large,..

A (couple of) spin-box(es) to set the outer (and inner) diameters might help there but you, Vadim, have previously reminded us to be really careful about adding more knobs and switches... :worried:

vadi2 commented:
Yeah, perhaps not on the general settings screen but hide the very rarely-used options behind pop-up menu. I don't foresee too many customising this given how few have asked to change it over the years.

SlySven commented:
Well I've added controls to set the size and also added alpha (transparency) controls to the custom colours which means that it can be adjusted to NOT obscure details - though what those might be can vary I guess between users' preferences for maps.

For instance grid mode maps perhaps will have different requirements than those with exits and where there is at least one (or two) room spaces between adjacent rooms - that is why I doubt there is one setting that everyone will think is the best...

vadi2 commented:
:+1:

SlySven commented:
It is still a Work-honestly-In-my-Plans-even-yesterday (**W**h**I**m**P**ey)... 😁 - it needs the settings to be save/loaded between session...

... also I would like to try and find a way to draw the player room last so then I can put it back to where it was before but avoid other rooms being painted from overwriting the new stuff. The quick and dirty fix would be to convert the `(QSet<int>) TArea::rooms` into a `QList<int>`  for each `T2DMap::paintEvent(...)` call and then, if the player room is found in that during the iteration through it to skip painting it then and to append it to the end of the list - this would probably be simpler but be less efficient code compared to refactoring out the whole of the *per room* paint code to an *inline* method and then calling it with the player room after the iteration through all the other rooms ...

vadi2 commented:
Hm... yes, let's just paint the room on top after instead of keeping changing the datastructures we use a lot.

SlySven commented:
Actually the refactoring isn't too bad and it does allow the `paintEvent` code to appear to be smaller (by shunting it off to an inlined {I hope} method - and the benefit of not overwriting the player room markings does seem worth it...

SlySven commented:
Oh, b\*\*\*\*r - I had two parallel repositories (so I can run side by side comparisons of my work in progress code against the main line one) both open in Qt Creator - and I did the refactoring **on the wrong repository's code**! :man_facepalming:

SlySven commented:
It turns out that using `auto` in a function prototype is a GCC extension to C++14 and the Clang compiler objects to it. (It was being used where a `QPen` reference/value was being passed to the new `inline T2DMap::drawRoom(...)`) It was inserted automagically by the latest Qt Creator *extract function* refactor operation so might need to be watched out for in the future...

vadi2 commented:
Good to know!

SlySven commented:
I'll see if I can fix it this morning - whilst on a 2 hour train to pick up a replacement :car: from a parent - I've reserved a Window seat with a table so should have power and WiFi on the way... :grinning:

vadi2 commented on src/T2DMap.cpp:
```diff
@@ -672,13 +672,503 @@ bool T2DMap::sizeFontToFitTextInRect( QFont & font, const QRectF & boundaryRect,
+// player's room and then, AFTER all those have been drawn, once for the
+// player's room if it is visible. This is so it is drawn LAST (and any effects,
+// or extra markings for it do not get overwritten by the drawing of the other
+// rooms)...
```
I don't think this comment is needed. We don't need the history of the code as a comment (we can look at git history for that) and we can tell the function is used in two places by looking at usages. Long comments don't really help in readability, plus it's only you who'd be interested in keeping it up to date...

SlySven replied:
Well, not every usage of the source code is going to be from a git repository, but I take your point - how about if I reduced it to a one-liner along the lines of:
```
// We need to ensure the player's room is drawn last so that any oversized parts of it don't get overwritten by other rooms....
```

vadi2 replied:
Perfect, and that should be at the point where this function is called for that reason - not at the function definition (doesn't make sense there)

vadi2 commented on src/T2DMap.cpp:
```diff
@@ -672,13 +672,503 @@ bool T2DMap::sizeFontToFitTextInRect( QFont & font, const QRectF & boundaryRect,
+    }
+    // We should be using the full area for testing for clicks even though
+    // we only show a smaller one if the user has dialed down the room size
+    // on NON-grid mode areas:
```
:+1:

vadi2 commented on src/T2DMap.cpp:
```diff
@@ -672,13 +672,503 @@ bool T2DMap::sizeFontToFitTextInRect( QFont & font, const QRectF & boundaryRect,
-    bool __Pick = mPick;
+    // As it happens this local is never changed in this method so it can be
+    // made const:
+    const bool __Pick = mPick;
```
```suggestion
    const bool pick = mPick;
```

SlySven replied:
Agreed! Whoever thought to using a double leading underscore is heading for trouble as far as I can see, e.g. https://stackoverflow.com/questions/224397/why-do-people-use-double-underscore-so-much-in-c - mind you, I think I have used them in the qmake project file but then that is not text that is parsed by the compiler/linker...

SlySven commented:
> @Eraene mind testing the final version?

Not the final version yet I am afraid - got to get "saving-to" and "loading-from" the game saves working and it is a bit awkward as the `TMap` (and the mapper classes) do not necessarily get instantiated when the the profile is being read.

*\<ASIDE>This means the settings have to be stored in the `Host` class whereas currently they are in the `TMap` one - furthermore, since we seem to be trying to protect the `Host` class from conflicting access from more than one thread with a `(QMutex) Host::mLock` it is desirable to not access that class's member variables more than is necessary {which is why I've just raised #2673 which suggests a way to reduce the risk of deadlock there} so ideally I should continue to use the copy of the settings in the `TMap` class for map *painting* but keep them synced with the master ones held in the `Host` class...\</ASIDE>*

SlySven commented:
Okay - **finally** got the save and load working.

Can you okay this now @vadi2 ?

*Helpful hint no. 42: using `pugixml::append_attribute(<attribute name>)` more than once on the same node with the **same*** attribute name *produces invalid XML - but as pugixml is a lightweight ("speedy"?) library it does not check for such unnatural acts {see https://github.com/zeux/pugixml/issues/269} and the Qt `QXmlStreamReader` class will only report that such an element with a duplicate attribute is an element of a `tokenType()` that is `QXmlStreamReader::Invalid` instead of `QXmlStreamReader::StartElement`.*

*The moral of this is - if adding a new attribute to one of the elements that are exported in an Mudlet `XMLexport` method ensure you do not forget to edit the attribute name if you copy and paste an existing `append_attribute(...)` line of code.* :scream:

vadi2 commented on src/T2DMap.cpp:
```diff
@@ -672,13 +672,503 @@ bool T2DMap::sizeFontToFitTextInRect( QFont & font, const QRectF & boundaryRect,
+// player's room if it is visible. This is so it is drawn LAST (and any effects,
+// or extra markings for it do not get overwritten by the drawing of the other
+// rooms)...
+inline void T2DMap::drawRoom(QPainter& painter, QFont& roomVNumFont, QPen& pen, TRoom* pRoom, TArea* pArea, const int currentRoomId, const bool isFontBigEnoughToShowRoomVnum, const int playerRoomId, const float rx, const float ry, const bool picked)
```
This function signature is a monster with many tentacles: I think if we gave anyone `drawRoom(painter, roomVNumFont, pen, room, pArea, currentAreaRoom, isFontBigEnoughToShowRoomVnum, playerRoomId, rx, ry, __Pick);` to look at, they'd have no clue what they were looking at. Let's improve it!

Can't `currentRoomId` be derived from `pRoom` inside the function? Or something like that - we have 3 arguments that all talk about the room, at least one should go.

`pArea` is only checked for `gridMode` - that should be passed in as a boolean to make the function interface more intuitive.

`isFontBigEnoughToShowRoomVnum` - no to yoda speak and **30** character variables! `roomTextfits` would work fine, observe:

```cpp
if (mShowRoomID && roomTextfits) {
```

And it wouldn't even need the comment it has right now, because it would be self-explanatory.

SlySven replied:
The many tentacles come from me using the refactor code to extract a large chunk of code from the `paintEvent` code but not spending enough time to refine out ***all*** the extraneous bits - the use of `pArea` to get the `gridmode` setting does sound sensible, getting `currentAreaRoom` from `pRoom` also; `roomTextFits` though - what text - there is a *symbol* is *that* the text it means - fair enough I will have a think on that and find something we both can agree on...

SlySven replied:
:worried: Oh, I haven't done much with this. Lemme have another look at it...

SlySven replied:
... okay take a look at the latest commit: https://github.com/Mudlet/Mudlet/pull/2621/commits/010f8b4776bca4cb2e68eca78c2f9589859845f7 !

vadi2 commented on src/T2DMap.cpp:
```diff
@@ -672,13 +672,503 @@ bool T2DMap::sizeFontToFitTextInRect( QFont & font, const QRectF & boundaryRect,
+            if (mpMap->mpRoomDB->getRoom(mTarget)) {
+                mpMap->mTargetID = mTarget;
+                if (mpMap->findPath(playerRoomId, mpMap->mTargetID)) {
+                    mpHost->startSpeedWalk();
```
Why is the function to draw a single room on a map calling `startSpeedWalk()`? That doesn't seem to make sense. Need to fix this in a follow-up PR.

SlySven replied:
:information_source: Nope, that ***is*** actually starting the speed walk route calculation - because the current room being painted has been determined to be close enough to where *the mouse was double-clicked* on the map to be the **destination** for the speed walk. Incidentally the use of the playerRoomId is because that is the starting point for the speed walk route.  Also, the destination room for the speedwalk gets drawn with a red-circle on it for just that one paint event.

SlySven replied:
As part of the previous point https://github.com/Mudlet/Mudlet/pull/2621/#discussion_r298793485 I happened to refactor this code out to a separate `initiateSpeedWalk(...)` helper! :angel:

vadi2 commented on src/T2DMap.cpp:
```diff
@@ -672,13 +672,503 @@ bool T2DMap::sizeFontToFitTextInRect( QFont & font, const QRectF & boundaryRect,
+                mTarget = it.key();
+                if (mpMap->mpRoomDB->getRoom(mTarget)) {
+                    mpMap->mTargetID = mTarget;
+                    if (mpMap->findPath(playerRoomId, mpMap->mTargetID)) {
```
Same

SlySven replied:
Similar as before - but this is because the *exit* from a room that goes to another area has been found to be close enough to the mouse double-click to be the destination for a speed-walk - also that exit will be drawn once (in this paintEvent only) with a red circle drawn on top of it...

vadi2 commented on src/TMap.cpp:
```diff
@@ -139,6 +139,15 @@ TMap::~TMap()
 
+void TMap::initialiseMembersFromHost()
+{
+    mPlayerRoomColorPrimary = mpHost->mPlayerRoomColorPrimary;
```
Why are we copying that into this class?

SlySven replied:
I originally wrote the code on the basis of having the settings in the `TMap` class - because they are associated with the map strangely enough :wink: - but then I worked out that they would need to be stored as part of the profile and thus as part of the `Host` class. I am aware that some of the members of that class are protected by a `QMutex` but it is not entirely clear why those particular ones are so protected; and rather than having to read the settings from that class for every `T2DMap::paintEvent()` I felt it was safer to work from copies held in the `TMap` class and only access the *master* ones in `Host` when they needed to be changed or used to initialise the copies.

vadi2 replied:
You're the one who put a `QMutex` on them... if we don't need it anymore, lets take it off?

Seems to make sense not to complicate things with forgotten mutexes and then copying things around...

vadi2 commented on src/dlgProfilePreferences.cpp:
```diff
@@ -3380,3 +3416,140 @@ void dlgProfilePreferences::slot_changeGuiLanguage(const QString &language)
+void dlgProfilePreferences::slot_setPlayerRoomSecondaryColor()
+{
+    Host* pHost = mpHost;
+    if (!pHost || !mpHost->mpMap || !mpHost->mpMap->mpMapper || !mpHost->mpMap->mpMapper->mp2dMap) {
```
This setting should be adjustable even when the map isn't open, right?

SlySven replied:
Debatable, I think. After all why show the controls if you are Mudding without a map - I know I wouldn't but not everyone thinks like that...

vadi2 replied:
Well - because you haven't yet opened the map?

SlySven replied:
Ah, looking at how the code is structured that slot directly alters the 2D mapper by calling a method in the `T2DMap` class - which can't happen if there is no map loaded. Going to have to rethink this...

vadi2 commented on src/dlgProfilePreferences.cpp:
```diff
@@ -3380,3 +3416,140 @@ void dlgProfilePreferences::slot_changeGuiLanguage(const QString &language)
+
+    auto color = QColorDialog::getColor(c, this, (b == pushButton_playerRoomPrimaryColor
+                                                  ? tr("Set outer color of player room marking.")
+                                                  : tr("Set innercolor of player room marking.")),
```
```suggestion
                                                  : tr("Set inner color of player room mark.")),
```

SlySven replied:
Yeah, makes sense.

SlySven replied:
Done in next commit.

vadi2 commented on src/dlgProfilePreferences.cpp:
```diff
@@ -3380,3 +3416,140 @@ void dlgProfilePreferences::slot_changeGuiLanguage(const QString &language)
+    }
+
+    auto color = QColorDialog::getColor(c, this, (b == pushButton_playerRoomPrimaryColor
+                                                  ? tr("Set outer color of player room marking.")
```
```suggestion
                                                  ? tr("Set outer color of player room mark.")
```

SlySven replied:
As above - yeah.

SlySven replied:
Done in next commit.

vadi2 commented on src/mudlet.cpp:
```diff
@@ -3297,6 +3297,9 @@ void mudlet::createMapper(bool loadDefaultMap)
     pHost->mpDockableMapWidget->setObjectName(QStringLiteral("dockMap_%1").arg(hostName));
+    // Values need to be copies across so they are in place when the 2D
+    // mapper is created:
+    pHost->mpMap->initialiseMembersFromHost();
```
I don't think we should change the design of how we read from the Host to this. It seems like more work and more error-prone. Just read when you need to and that's it...

SlySven replied:
I will have to check but I *think* I may use the values in the `TMap` in the constructor for the `T2DMap` instance (which is created as part of the construction of the `dlgMapper` class)...

... actually I realise I may not have done the necessary in the other place where that construction happens - for the Mapper embedded in the main `TConsole` - :worried:

vadi2 replied:
Right, see - this way of setting values is error-prone. Let's not design like this, let's have stick to the simpler and more robust design we already have (read from the host when you need to, cache at the start of the method if that's necessary).

SlySven commented on src/Host.h:
```diff
@@ -426,6 +426,25 @@ class Host : public QObject
+    QColor mPlayerRoomColorPrimary;
+    QColor mPlayerRoomColorSecondary;
+    // Mode selected - 0 is closest to original style:
+    quint8 mPlayerRoomStyle;
```
Deliberate limit on something with less than a handful of choices (four), and it is much safer to convert between integer representations of numbers of different bitnesses for unsigned integers.

vadi2 replied:
This seems to go exactly against the best practice mentioned above?

SlySven commented on src/Host.h:
```diff
@@ -426,6 +426,25 @@ class Host : public QObject
     // An invalid/null value is treated as the "show all"/inactive case:
     QTime mTimerDebugOutputSuppressionInterval;
 
+    // This hold copies of variables from the TMap class which are saved with
```
It didn't evolve like that! When I started this PR I hadn't contemplated the load/save of the settings - though in hindsight that comment does need revision..

vadi2 commented:
@SlySven ping

vadi2 commented:
:wave: :)
